youtube_dl/extractor/brightcove.py

```diff
@@ -501,22 +504,39 @@ def _extract_urls(webpage):
                 r'<iframe[^>]+src=(["\'])((?:https?:)?//players\.brightcove\.net/\d+/[^/]+/index\.html.+?)\1', webpage):
             entries.append(url if url.startswith('http') else 'http:' + url)
 
-        # Look for embed_in_page embeds [2]
-        for video_id, account_id, player_id, embed in re.findall(
-                # According to examples from [3] it's unclear whether video id
-                # may be optional and what to do when it is
-                # According to [4] data-video-id may be prefixed with ref:
-                r'''(?sx)
-                    <video[^>]+
-                        data-video-id=["\'](\d+|ref:[^"\']+)["\'][^>]*>.*?
-                    </video>.*?
-                    <script[^>]+
-                        src=["\'](?:https?:)?//players\.brightcove\.net/
-                        (\d+)/([^/]+)_([^/]+)/index(?:\.min)?\.js
-                ''', webpage):
-            entries.append(
-                'http://players.brightcove.net/%s/%s_%s/index.html?videoId=%s'
-                % (account_id, player_id, embed, video_id))
+        # Look for <video> tags [2] and embed_in_page embeds [3]
+        # [2] looks like:
+        # <video data-video-id="5320421710001" data-account="245991542" data-player="SJWAiyYWg" data-embed="default" class="video-js" controls itemscope itemtype="http://schema.org/VideoObject">
+
+        for video, script_tag, account_id, player_id, embed in re.findall(
+            r'''(?isx)
+                (<video[^>]+>)
+                (?:.*?
+                  (<script[^>]+
+                    src=["\'](?:https?:)?//players\.brightcove\.net/
+                    (\d+)/([^/]+)_([^/]+)/index(?:\.min)?\.js
+                  )
+                )?
```

Comment:
> Do you have example URLs with video tags only and without brightcove indication? 

This is a good point. I have two examples that have what I would call bad "indications" and I'm not sure how they qualify. From #9163 there's  http://ew.com/tv/2017/03/08/guardians-galaxy-cartoon-exclusive-clip/ which looks like this (adding line breaks):

```html
<!-- Start of Brightcove Player -->
<div class="video__wrapper">
<video
  data-account="219646971" data-player="default"
  data-video-id="5352121610001" data-embed="default" class="video-js"
  data-autoplay="false" data-viewplay="true" data-stickyplay="true"
  controls playsinline style="width: 100%; height: 100%;">
</video>
</div>
<!-- End of Brightcove Player -->
```

so the only indication is the comments surrounding it.

And also #12005 gives us http://www.programme-tv.net/videos/extraits/81095-guillaume-canet-evoque-les-rumeurs-d-infidelite-de-marion-cotillard-avec-brad-pitt-dans-vivement-dimanche/ which has lines 41-46:

```html
    	    <script>
		$script([
			'//players.brightcove.net/800000640001/HyenP6LuN_default/index.min.js',
			'//players.brightcove.net/800000640001/Hyx5mvd4_default/index.min.js'
		], 'brightcove')
	</script>
```

and then the video itself around line 467:

```html
<div class="video_partenaire id_brightcove_5311302538001 d-b p-r maxw-100">
    <div class="ratio-16-9">
        <video data-video-id="5311302538001"
               data-account="800000640001"
               data-player="HyenP6LuN"
               data-embed="default"
               class="video-js apply-ratio"
               autoplay preload controls>
	</video>
    </div>
</div>
```
So there's the `<script>` that doesn't fit the standard pattern (no `src=` attribute), although it's close, modulo the JQuery or whatever `$script` is. And the `id_brightcove_5311302538001` in the `class` attribute of the grandparent `<div>` . Sketchy.

And also there's the `[Bostonglobe]` case from #12099 which you don't consider appropriate for this extractor but we should at least think about, which has brightcove refences but no script reference (the script references are buried in javascript loaded by javascript loaded by javascript):

```html
<div class="videoplayer" id="video">
  <video data-brightcove-video-id="5320421710001" data-account="245991542" data-player="SJWAiyYWg" data-embed="default" class="video-js" controls itemscope itemtype="http://schema.org/VideoObject">
    <meta itemprop="name" content="A tree finally succumbs to disease, leaving a hole in a neighborhood" />
    <meta itemprop="thumbnailURL" content="http://www.bostonglobe.com/rf/image_130x80/Boston/2011-2020/2017/02/12/BostonGlobe.com/Metro/Videos/Sources/5320421710001.jpg" />
    <meta itemprop="contentURL" content="https://secure.brightcove.com/services/mobile/streaming/index/master.m3u8?videoId=5320421710001&amp;pubId=245991542&amp;secure=true" />
    <meta itemprop="width" content="867" />
    <meta itemprop="height" content="486" />
  </video>
</div>
```
That is identifyable both by the `data-brightcove-video-id` attribute and also the presence of the `<meta itemprop="contentURL">` that with a Brightcove URL (Huh, hadn't noticed that before).

So, do you consider any of the above indications sufficient to check for?

Maybe the right answer is to move the search for `<video>`  tags into `generic.py`, and then call a series of methods from extractors (like `._extract_urls()` but different) to determine if they want to try to retrieve (perhaps the Brightcove methods would return `Yes` in cases where the standard `<script ... players.brightcove.net>` was present, and `Maybe` in the other cases. Then `GenericIE` could call the `Yes` extractor(s) if there is one (any), or iterate over the `Maybe`s if not.

This also gets tricky if we allow for the possibility of multiple videos on a single page, esp. from different embedding sites.

In practice, does youtube-dl support any other extractors for `<video>` tags with this kind of ambiguity, and are any of them likely embedable in different domains like Brightcove is?

`CrackedIE` has a straight-up `<video src="/path/to/file.mp4">`, as does `VimeoIE`; `LifeNewsIE` has some similar support though the tests don't include those cases, `PornHubIE`, and `XHamsterIE` too. `TwitterIE` deals with `<video class="animated-gif">`, although I didn't find a test that used it.

`LibraryOfCongressIE` has a `data-download` (though the extractor seems not to use it, relying on a uuid?). `SVTPlayIE` uses `data-video-id`. 

`VideoMegaIE` uses it to get a thumbnail, only.

Oh, and I guess `_parse_html5_media_entries()` also deals with `<video>` tags, but I grepped for `'<video'` and that missed `r'(?s)(<(?P<tag>video|audio)(`,. I'm not quite sure of the applicability of that...

Oh, also `RTSIE`, maybe?